Add/nodeipamcontroller

DockToFuture opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
 /kind feature


**What this PR does / why we need it**:
This change adds a `nodeipam controller` to the aws cloud-controller-manager for IPv6 prefix delegation. The node ipam controller assigns POD addresses to kubernetes nodes based on prefix delegations. The controller supports plain IPv6 (`--dualstack=false`) and dual stack (`--dualstack=true`). In dual stack mode the IPAM for IPv4 is done in the same way it was done in the kube-controller-manager.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/cloud-provider-aws/issues/608

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
This change adds a `nodeipam controller` to the aws cloud-controller-manager for IPv6 prefix delegation. The node ipam controller assigns POD addresses to kubernetes nodes based on prefix delegations. The controller supports plain IPv6 (`--dualstack=false`) mode and dual stack (`--dualstack=true`) mode. 
```

cartermckinnon commented:
/assign @cartermckinnon

cartermckinnon commented:
/assign @M00nF1sh

hakman commented:
/ok-to-test

dims commented:
/test all

dims commented:
/kind feature
/triage accepted

dims commented:
@DockToFuture how did you test this? I can't spot any tests in this PR, are parts of this controller here borrowed from other places? could we borrow the tests that exist there as well?

DockToFuture commented:
Hello @dims I've deployed the controller into our IPv6 Kubernetes clusters and test it with the following controllers(`--controllers=cloud-node,cloud-node-lifecycle,nodeipam,service,route`).
The code is similar to https://github.com/kubernetes/cloud-provider-gcp/blob/master/cmd/cloud-controller-manager/nodeipamcontroller.go . What for tests do you expect? Can you elaborate on it a little bit?

dims commented:
@DockToFuture so what i was thinking was if we were to enhance one of the existing harness(es) to enable this control, for example `make test-e2e` uses `kops` to setup a cluster and run a bunch of tests
- https://github.com/kubernetes/cloud-provider-aws/blob/master/Makefile#L177-L184
- https://github.com/kubernetes/cloud-provider-aws/blob/master/hack/e2e/run.sh#L124-L134

if we take this harness (switch on ipv6!), add a way to ensure that this new controller is activated/enabled that would be the first step. then we could add some basic golang based tests in test/e2e folder to ensure we don't break this controller going forward. You called out `The node ipam controller assigns POD addresses to kubernetes nodes based on prefix delegations` in your PR body as the feature, so that's what we would test, launch pods and test if the ip addresses look right.

Does that make sense?

dims commented:
@DockToFuture also there are some unit tests there in that other repo you pointed to, some of it may be ported over here as well? 
https://cs.k8s.io/?q=package&i=nope&files=.*nodeipam.*_test%5C.go&excludeFiles=&repos=kubernetes/cloud-provider-gcp

olemarkus commented:
It should be possible to use this as a drop-in replacement to the ipam controller we implemented in kOps. 
I expect kOps would still handle assigning the actual prefixes to instances.

kOps would not support dualstack though, so it may be an idea to mark that option experimental unless there is another way of getting that e2e-tested.

DockToFuture commented:
> @DockToFuture also there are some unit tests there in that other repo you pointed to, some of it may be ported over here as well? https://cs.k8s.io/?q=func%20Test&i=nope&files=.*nodeipam.*_test%5C.go&excludeFiles=&repos=kubernetes/cloud-provider-gcp

I've added all relevant unit tests and adapted them where necessary. I've also added some unit tests for the new functionality.

DockToFuture commented:
> It should be possible to use this as a drop-in replacement to the ipam controller we implemented in kOps. I expect kOps would still handle assigning the actual prefixes to instances.
> 
> kOps would not support dualstack though, so it may be an idea to mark that option experimental unless there is another way of getting that e2e-tested.

I've marked the flag as experimental.

DockToFuture commented:
> * https://github.com/kubernetes/cloud-provider-aws/blob/master/Makefile#L177-L184

What is the best way to do that? Create a cluster with kops and afterwards edit the deployment of the cloud-controller-manager in the e2e test or is there a way to directly set the flags for the cloud-controller-manager in the kops cluster creation call? I didn't find any information how this could be done. Any ideas?

DockToFuture commented:
> @DockToFuture so what i was thinking was if we were to enhance one of the existing harness(es) to enable this control, for example `make test-e2e` uses `kops` to setup a cluster and run a bunch of tests
> 
> * https://github.com/kubernetes/cloud-provider-aws/blob/master/Makefile#L177-L184
> * https://github.com/kubernetes/cloud-provider-aws/blob/master/hack/e2e/run.sh#L124-L134
> 
> if we take this harness (switch on ipv6!), add a way to ensure that this new controller is activated/enabled that would be the first step. then we could add some basic golang based tests in test/e2e folder to ensure we don't break this controller going forward. You called out `The node ipam controller assigns POD addresses to kubernetes nodes based on prefix delegations` in your PR body as the feature, so that's what we would test, launch pods and test if the ip addresses look right.
> 
> Does that make sense?


What is the best way to do that? Create a cluster with kops and afterwards edit the deployment of the cloud-controller-manager in the e2e test or is there a way to directly set the flags for the cloud-controller-manager in the kops cluster creation call? I didn't find any information how this could be done. Any ideas?

DockToFuture commented:
When creating a kops cluster with `--ipv6` flag, the networking CIDR is set automatically to the IPv4 CIDR of the VPC`172.20.0.0/16`.
This leads to worker nodes not joining the cluster because IPv4 addresses are none routable in the private IPv6 subnets and thus the bootstrapping fails.
```
Dec 08 13:08:45 i-xxxxxxxxxxxxxxx nodeup[1347]: W1208 13:08:45.148005    1347 main.go:133] got error running nodeup (will retry in 30s): failed to get node config from server: Post "https://172.20.1.200:3988/bootstrap": dial tcp 172.20.1.200:3988: connect: network is unreachable; Post "https://172.20.6.164:3988/bootstrap": dial tcp 172.20.6.164:3988: connect: network is unreachable; Post "https://172.20.9.26:3988/bootstrap": dial tcp 172.20.9.26:3988: connect: network is unreachable
```
When editing the cluster config and setting it to the IPv6 CIDR of the VPC, kops throws an error:
```unexpected IP address type: xxxx:xxxx:xxxx:xx00::/56```.
How can I get kops to use the correct IPv6 CIDR range of the VPC? Did this even ever work?

DockToFuture commented:
I've update the PR with an e2e test. A valid domain needs to be provided to run the e2e tests successfully. It's required to get the kops cluster up and running as the used gossip implementation does not support IPv6 yet. It can be set via`DNS_ZONE` environment variable. I've used a route53 hosted zone domain for testing.

DockToFuture commented:
PR is now ready for review! :)

M00nF1sh commented on hack/e2e/run.sh:
```diff
@@ -143,10 +145,62 @@ fi
 set -x
 pushd ./tests/e2e
-ginkgo . -v -p --nodes="${GINKGO_NODES}" --focus="${GINKGO_FOCUS}" --skip="${GINKGO_SKIP}" --report-dir="${ARTIFACTS}"
+ginkgo --focus="${GINKGO_FOCUS}" --skip="${GINKGO_SKIP}" --label-filter="${GINKGO_LABEL_FILTER}" . -v -p --nodes="${GINKGO_NODES}"  --report-dir="${ARTIFACTS}"
```
why we added a filter for loadbalancer here

DockToFuture replied:
Because I only want to run the `loadbalancer` test here and not the `ipv6 prefix` test. In general we could run both tests here and they will succeed but we will get nothing out of it.

M00nF1sh replied:
seems we now have a node test suit: https://github.com/kubernetes/cloud-provider-aws/blob/master/tests/e2e/nodes.go

this will be a regression that we skipped the "nodes" test suit when run e2e.
i think we shouldn't have this label-filter, and should run all tests we have.

M00nF1sh commented on pkg/controllers/nodeipam/config/types.go:
```diff
@@ -0,0 +1,30 @@
+
+// NodeIPAMControllerConfiguration contains elements describing NodeIPAMController.
+type NodeIPAMControllerConfiguration struct {
+	RateLimit    float64
```
we need some documentation & better name for those fields.
e.g. what's the `RateLimit` for? could there be other "RateLimit" in the future than this one that need to be configured

DockToFuture commented:
Any more feedback @M00nF1sh?

dims commented:
@DockToFuture please rebase and i'll merge. thanks!

DockToFuture commented:
Hello @dims, I've rebased the PR.

dims commented:
looks like failure in `pull-cloud-provider-aws-e2e` is at cleanup phase and the test passed

```
[38;5;10m[1mSUCCESS![0m -- [38;5;10m[1m1 Passed[0m | [38;5;9m[1m0 Failed[0m | [38;5;11m[1m0 Pending[0m | [38;5;14m[1m1 Skipped[0m
PASS

Ginkgo ran 1 suite in 1m28.931164668s
Test Suite Passed
```

dims commented:
/skip

dims commented:
/approve
/lgtm

dims commented:
/retest

DockToFuture commented:
@dims a valid dns-zone needs to be set in the environment variable "DNS_ZONE" in order to create the kops cluster successfully, see my earlier comment: https://github.com/kubernetes/cloud-provider-aws/pull/679#issuecomment-1881073904 for more information.

```
Error: cannot find DNS Zone "example.com".  Please pre-create the zone and set up NS records so that it resolves
Error: exit status 1
make: *** [Makefile:182: test-e2e] Error 1
```

dims commented:
@M00nF1sh any suggestions on how to make this work?

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+	// cidrSet are mapped to clusterCIDR by index
+	cidrSets := make([]*cidrset.CidrSet, len(allocatorParams.ClusterCIDRs))
+	for idx, cidr := range allocatorParams.ClusterCIDRs {
+		cidrSet, err := cidrset.NewCIDRSet(cidr, allocatorParams.NodeCIDRMaskSizes[idx])
```
nit, this can cause crash if the len(allocatorParams.NodeCIDRMaskSizes) <  len(allocatorParams.ClusterCIDRs).
It's a nit comment given this shall be supplied by cluster admins so lt's low risk, but good to have validations with clear error message when len(allocatorParams.NodeCIDRMaskSizes) != len(allocatorParams.ClusterCIDRs)

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
```
nit, we can optimize the performance and readability by doing the check once before the loop
e.g. 
```
if len(node.Spec.PodCIDRs) != len(r.cidrSets) {
   return "some error"
}
for idx, cidr := range node.Spec.PodCIDRs {
 .....original logic without the idx >= len(r.cidrSets) check
}
```

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+		// If node has a pre allocate cidr that does not exist in our cidrs.
+		// This will happen if cluster went from dualstack(multi cidrs) to non-dualstack
+		// then we have now way of locking it
+		if idx >= len(r.cidrSets) {
```
same as previous comment, i'd like to move this check before the loop for performance and readability.
the len(node.spec.PodCIDRs)(when it's not 0) must always **equal to**(instead of `<=`)  len(r.cidrSets)

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+// function you have to make sure to update nodesInProcessing properly with the
+// disposition of the node when the work is done.
+func (r *rangeAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
+	if node == nil {
```
just curious, when will node been nil? or is this sort of Defensive programming(if so, let's add a comment like this should never happen for readability) or remove this check if we can ensure it's non-nil.

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+
+	for idx := range r.cidrSets {
+		podCIDR, err := r.cidrSets[idx].AllocateNext()
+		if err != nil {
```
if this happens, seems the allocations for other index(< current idx) will be leaked

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+		}
+
+		klog.V(4).Infof("release CIDR %s for node:%v", cidr, node.Name)
+		if err = r.cidrSets[idx].Release(podCIDR); err != nil {
```
shall we do a best effort by try release for each cidrSet and return aggregated error?

M00nF1sh commented on pkg/controllers/nodeipam/ipam/cidr_allocator.go:
```diff
@@ -0,0 +1,480 @@
+	}
+
+	// aws:///eu-central-1a/i-07577a7bcf3e576f2
+	instanceID, _ := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
```
the error is ignored..what id the instance id is a non-valid instanceid? like a fargate task?

M00nF1sh commented on pkg/controllers/nodeipam/ipam/ipv6_allocator.go:
```diff
@@ -0,0 +1,284 @@
+	}
+
+	// aws:///eu-central-1a/i-07577a7bcf3e576f2
+	instanceID, _ := awsv1.KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
```
the error is ignored.

M00nF1sh commented on pkg/controllers/nodeipam/nodeipam_controller.go:
```diff
@@ -0,0 +1,218 @@
+			recorder:              recorder,
+			nodesInProcessing:     sets.NewString(),
+		}
+		nodeList, err := listNodes(kubeClient)
```
why we need to use listNodes? shouldn't the informer handles all existing nodes?

M00nF1sh reviewed (commented):
Another high level comment:
1. why we choose to have different implementation pattern for the "dualstack" and "ipv6only" allocator? Having different patterns is hard to maintain since technically they are doing the same work..
  * with the dualstack one, the major work is handled directly during informer's event handler, while uses a worker queue to defer the cidr updates. 
  * with the ipv6only one, the informer's event handler merely distribute the node event into a queue, and workers then handles it.


2. the cidr_allocator seems to shall be named like dualstack_allocator(or more specifically cidr_plus_ipv6prefix_allocator). it's not only allocate from preconfigured list of cidrs but also picks up allocates from existing IPv6 prefixes on node). I'm confused by the use case, why would people use both ip from configured list & ipv6 prefixes together?

DockToFuture commented:
I will close this PR as we have decided to implement an own IPAM controller as suggested by @M00nF1sh to keep the CloudProvider AWS as slim as possible (analogous to the AWS [loadbalancer controller](https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.7/)). Furthermore it gives us more degrees of freedom how the tests for the IPAM controller are run and executed.

dims commented:
thanks @DockToFuture !
